Use useMemo for hydration and avoid unnecessary hydrates

voinik opened the pull request:
## **Server side problems**
First of all, I've solved the use of double `useHybridHydration` hook calls. See my comment here: https://github.com/kirill-konshin/next-redux-wrapper/pull/499/files#r1014500941 (thanks to @karaoak).
~Second, I fixed the need for server side hydration. I did it by sharing the same reference to the store in the `context.req.__nextReduxWrapperStore` and the non-`context.req` flow.~ ⬅️ This is **BAD**, **DO NOT DO THIS**. See my comment here: https://github.com/kirill-konshin/next-redux-wrapper/pull/502#discussion_r1031531425.

## **Client side problems**
These problems arise when doing **client side navigation**. So keep in mind that that's what I'm discussing here.

First keep in mind how the data flows through Next.js at the moment. See my comment here: https://github.com/kirill-konshin/next-redux-wrapper/pull/499/files#r1014500941

After thinking about the issues in #493,  #495 and #496, I've come up with 2 possible solutions, both with some caveats.

~### Solution **A**~
~Instead of `useEffect` we use `useLayoutEffect`. It runs **before** `useEffect`, so issues relating to children of the consuming component changing the store in their `useEffects` should be resolved with this approach.~
~The flow in this solution goes like this:~
~1. Data from the server flows into the component in `_app.tsx`; let's call it the `App` component~
~2. `App` starts to rerender~
~3. Data is passed to `useWrappedStore`~
~4. We schedule the hydration **after** all child renders/mounts, but **before** any paints (that's what `useLayoutEffect` does~
~5. `App` continues to rerender and hits the return, and mounts and renders its children for the new route we navigated to~
~6. Mounting and rendering is complete, but we have not painted yet, so the user doesn't see anything new yet~
~7. We hydrate~
~8. Data changes in 2 important places:~
    ~a. The store changes in the `App` component, so the `store` variable in the `const {store, props} = wrapper.useWrappedStore(rest);` line triggers a rerender of the `App` component~
    ~b. The newly rendered child component for the new route (step 6) has selectors in it that depend on the store. This change triggers a rerender for this child.~
~9. This triggers a race condition. The `App` component is trying to rerender at the same time as the newly rendered child, which triggers a warning. You'll see this warning rarely (usually only once when you refresh, and then do a client side navigation), but it's there.~
~10. React resolves this and renders the `App` and child properly, and then finally paints with the correct data.~
~11. The user doesn't see any layout shift, or anything appear/disappear~

~**The caveats**~
~* In step 6, the children for the new route are mounted and rendered. However, if they use selectors, then those selectors will be derived from whatever the state is **BEFORE** hydration for this new page. So if your selectors have nested properties like `state.foo.bar`, and your state **BEFORE** hydration of this new page is something like `state = null`, then you'll get null pointer exceptions (e.g. cannot read `foo.bar` from `null`). This means all users MUST use optional chaining (`state?.foo.bar`) or their new pages will crash.~
~* The fact that the selector can be `undefined` should not be a problem because the hydration occurs **before** the paint.~
~* If, however, your selector returns an object, and your jsx tries to access nested properties there, you'll have to do the chaining there too.~

### Solution **B**
Instead of `useEffect` we use `useMemo`. It runs before **all** effects; in fact it runs **synchronously,** so issues relating to children of the consuming component changing the store in their `useEffects` should be resolved with this approach as well.
The flow in this solution goes like this: 
1. Data from the server flows into the component in `_app.tsx`; let's call it the `App` component
2. `App` starts to rerender
3. Data is passed to `useWrappedStore`
4. We do the hydration **now**, so before any children are mounted/rerendered
5. ~`App` notices the `store` variable has changed, so a new rerender is scheduled~ The hydration gives us our new store which is subsequently used in the rest of the rerender. This does not count as a mutation because it's synchronous, so no extra rerender is needed.
6. `App` finishes the ~first~ rerender, thereby hitting the return, and mounts and renders its children for the new route we navigated to
7. Mounting and rendering is complete, and we paint everything with the correct data
8. ~`App` finishes the second rerender (triggered in step 5; again, not sure if this is how the scheduling works)~ 

**The caveats**
* ~Between step 4 and 5 this is still the case: **The currently mounted children are still there!**~
* ~This means that the hydration from step 4 has changed the selectors of the currently mounted components~
* ~If your hydration slices do not do state reconciliation (i.e. they don't save the current state but overwrite the entire state with the incoming state), then if your selectors use nested properties and your initial state is not chainable, you'll get null pointer exceptions. The reason is the server always rebuilds the store on every request for the slice of the new page. The other slices in your store will use the initial state. So when you hydrate and you don't do anything to save the current state (state reconciliation), your selectors in your current component will suddenly use your initial state.~
* ~This is fixable by doing optional chaining here too, but there's a chance the user may see data disappearing, because the data the currently mounted children depend on is suddenly gone, before any re-mounts or rerenders have happened.~
* We're blocking the react lifecycle loop of the `App` component by using `useMemo` for all hydrations. I think this is more of an anti-pattern than `useLayoutEffect`. Effects should be done after mounting, and hydrations are still an effect. This flow feels more like an effect than `useMemo`.
* We get a warning that we're trying to rerender a component while the `App` rerender is happening (because the selector changed in the old page component).

The above struckthrough caveats are not applicable anymore. Turns out that when doing client side navigation, React does **not** actually rerender when the selectors in the (still mounted) previous page component change. See my comment in `useHybridHydrate` in `/packages/wrapper/src/index.tsx`. The selector does run, but because the there is no rerender of the old component, no null pointer exceptions will be thrown, as the code that relies on the data being there is not run. There is a warning that a rerender is triggered, but it's not actually run.

## Conclusion
~I personally think solution **A** is best because it sticks to the react best practices the most of the two (even though using `useLayoutEffect` is still an anti-pattern). That is why I've implemented that solution in this PR.~
I now think solution **B** is the best, because it doesn't require consumers of this package to do optional chaining nor data reconciliation. Users won't have to do anything to use this new approach. So even though we're blocking React, I think it's still worth it from an OSS perspective. Lastly, we still have the warning that we're rerendering 2 components at once (`App` and the old page component we're navigating away from), but this is the case in both approaches, and I haven't been able to find a solution that avoids this.

If you'd like to see solution **A**, check out the `useHybridHydrate` function in `/packages/wrapper/src/index.tsx` using this command: `git show 32c683ea1addebf9ae968a1077abe982dec3d718` (or find that commit in the commit history on GitHub).

Let me know what you think!

kirill-konshin commented on packages/wrapper/src/index.tsx:
```diff
@@ -55,12 +60,16 @@ const initStore = <S extends Store>({makeStore, context = {}}: InitStoreOptions<
 
-        return createStore();
+        // This allows us to have the same store that we modified in the GIP/GSSP/GSP dispatches without having context available
+        return serverStore || createStore();
```
Does `serverStore` remain between requests?

voinik replied:
No, the file is loaded all over again on each request server side. So it doesn’t persist

voinik replied:
I figured I could explain why it works.

The reason this fixes the hydration requirement is because in GIP and GSSP there is a context object that we get from next, and we store the state in it (`context.req.__nextReduxWrapperStore`). But when the data flows into the `App` component, so when we get into React, we no longer have a context object, so we can't reach the `__nextReduxWrapperStore` property anymore.

Because this file is still used in the server runtime during GIP/GSSP and during the server render of `App`, we can put a reference to the store in a separate variable; the `serverStore` variable. It will be alive during GIP/GSSP **and** during the `App` render. 

Because the store is a reference that's shared between `__nextReduxWrapperStore` and `serverStore`, any modifications to the store will be accessible to both. So GIP/GSSP use the `__nextReduxWrapperStore`, and the `useWrappedStore` hook in `App` uses `serverStore` to reach the same store.

kirill-konshin replied:
Well, in this case we can omit using `context.req` and always use module variable instead. In any case this is not even a public API, so it's a safe change.

kirill-konshin commented on packages/wrapper/src/index.tsx:
```diff
@@ -161,26 +170,40 @@ export const createWrapper = <S extends Store>(makeStore: MakeStore<S>, config:
-            if (getIsServer() || firstRender.current) {
+            // useMemo runs on the server, but we should not hydrate on the server...
+            if (getIsServer()) {
+                if (config.serializeState) {
```
Not sure why this is needed here

kirill-konshin commented:
First of all, let me thank you for the great job, very thorough explanation.

Maybe we don't need to bump the major version, as the public APIs remain the same.

kirill-konshin commented on packages/demo-redux-toolkit/pages/subject/[id].tsx:
```diff
@@ -1,42 +1,63 @@
+    console[pageName ? 'info' : 'warn']('Rendered pageName: ', pageName);
+
+    if (!pageName) {
+        // On client side routing from the homepage to this route, the selectors will be undefined because of the
```
I suggest to add `throw new Error` here so that it will be more obvious that going in this if is errornous.

kirill-konshin replied:
In my experiments, I've tried to use `useMemo` to synchronously dispatch, https://github.com/kirill-konshin/next-redux-wrapper/blob/async-hydration-fix/packages/wrapper/src/index.tsx#L175 and it kinda worked.

I used layout effect only for cases when pathname did not change. This addresses the known problems.

voinik replied:
I've addressed it, check out the edited PR description and my other comments

kirill-konshin commented on packages/demo-redux-toolkit/store.tsx:
```diff
@@ -1,58 +0,0 @@
-
-export const wrapper = createWrapper<AppStore>(makeStore);
-
-export const selectSubject = (id: any) => (state: AppState) => state?.[subjectSlice.name]?.[id];
```
I purposely had everything in one file to make it simpler to consume as example, multiple files are OK for real use.

kirill-konshin commented on packages/demo/tests/index.spec.ts:
```diff
@@ -55,7 +55,7 @@ test('other page -> static', async ({page}) => {
     await page.click('text=Navigate to static');
 
-    await expect(page.locator('body')).toContainText('"page": "static"'); // redux
+    await expect(page.locator('body')).toContainText('"page": "server"'); // cannot do dispatches in getStaticProps -> previous state
```
Static pages still should return some state, and hydrate can decide whether to put it or not. If user visits GSSP page first and then visits the GSP page, the GSP state has to be delivered.

voinik replied:
I've reverted these changes, but I would still like to know why this is happening 😕

kirill-konshin replied:
Well, you can put some stuff in the store during the build, and it will be preserved in static props. So when you open the page, these props will be used for hydration.

kirill-konshin commented on packages/wrapper/src/index.tsx:
```diff
@@ -197,8 +220,9 @@ export const createWrapper = <S extends Store>(makeStore: MakeStore<S>, config:
-        useHybridHydrate(store, initialStateFromGSPorGSSR);
+        // If GSSP/GSP/GIP has run, then pageProps.initialState contains the most complete data
+        // If it hasn't, then we use the initial state (see https://github.com/kirill-konshin/next-redux-wrapper/pull/499/files#r1014500941)
+        useHydrate(store, initialStateFromGSPorGSSR ?? initialState ?? null);
```
Great explanation in your comment. Let's keep this new single hydrate

kirill-konshin commented:
Please fix some minor things that I pointed out, and I will merge this and release. This PR so far is the best candidate to be merged. 

Looks like you already have taken necessary code from my PR #498. Maybe also take Pokemon [page](https://github.com/kirill-konshin/next-redux-wrapper/pull/498/files#diff-0f2ce52a407a8db7948e4a83e148d686320a200c688ff1799d5529c7997d5cc1) and [API](https://github.com/kirill-konshin/next-redux-wrapper/pull/498/files#diff-ee9392c4e05cc38a8e69f0b6d81e294490f5bc1ced1809b39785d4d8131c942a) as demonstration of Redux Toolkit API instead of RT Slices.

voinik reviewed (commented):
Alright I believe I've covered everything. I have some questions about the GSP that I posted in this PR, but other than that it seems we're good.

I would like to one more time bring your attention to the fact that the `useLayoutEffect` approach may cause crashes if people have not used optional chaining in their selectors, because the rerender occurs **before** the hydrate. So if there's no optional chaining, people might get null pointer exceptions.
However, users won't see the bad data because the hydrate occurs **before** the paint.

The second caveat is that the warning I mentioned in the first post for solution **A** may appear (only once per refresh though), but not in production.

If you're okay with both of these caveats then I think we're good to go.

Let me know if you have more feedback that I can incorporate before the merge 😁 

_Edit:_ I forgot to mention: These caveats already exist on master, because the current implementation uses `useEffect`, except it's worse than `useLayoutEffect` because the hydrate is done after the paint, so users might see bad data. My point is that the caveats I mention already exist on master

kirill-konshin commented on packages/wrapper/src/index.tsx:
```diff
@@ -161,27 +163,34 @@ export const createWrapper = <S extends Store>(makeStore: MakeStore<S>, config:
+
+              // useMemo so that the very first hydration runs synchronously, before any child renders
+              useBrowserMemo(() => {
+                  if (firstHydrate.current) {
```
As you can see in my experiments https://github.com/kirill-konshin/next-redux-wrapper/blob/async-hydration-fix/packages/wrapper/src/index.tsx#L175 I use not first hydrate, but new page detection here. And it seemed to work...

voinik commented on packages/wrapper/src/index.tsx:
```diff
@@ -161,44 +161,59 @@ export const createWrapper = <S extends Store>(makeStore: MakeStore<S>, config:
+        const newPage = prevRoute.current !== asPath;
+        prevRoute.current = asPath;
+
+        // useMemo so that when we navigate client side, we always synchronously hydrate the state before the new page
```
You (@kirill-konshin) were right in one of the previous comments you had. Using `useMemo` for hydration (solution **B**) will hydrate **before** the new page component is mounted, that is, while the old page component is still mounted. But it turns out React **doesn't** trigger a rerender on the currently mounted page component, even though the selectors it uses change (see comment under the `detailPageInitialState` definition in `/packages/demo-redux-toolkit/store.ts`).

This is great, because that means optional chaining is no longer required, and data reconciliation is also not needed for this approach (solution **B**). In other words, users of this library will not have to change anything in their code. This is in contrast to solution **A**, which does require optional chaining.

voinik commented on packages/wrapper/src/index.tsx:
```diff
@@ -161,44 +161,57 @@ export const createWrapper = <S extends Store>(makeStore: MakeStore<S>, config:
-        // this happens when App has page with getServerSideProps/getStaticProps, initialState will be dumped twice:
-        // one incomplete and one complete
-        const initialStateFromGSPorGSSR = props?.pageProps?.initialState;
+        const gspState = props?.__N_SSG ? props?.pageProps?.initialState : null;
```
Next provides these props to tell us what type of server side data fetching is happening. We do different things based on which one it is in the orchestrator. I feel like this approach is more clear to show what's going on in the data flows and which state (`gspState` and `gsspState`) is used where.

voinik commented on packages/wrapper/src/index.tsx:
```diff
@@ -55,15 +56,14 @@ const initStore = <S extends Store>({makeStore, context = {}}: InitStoreOptions<
             // @see https://github.com/kirill-konshin/next-redux-wrapper/pull/196#issuecomment-611673546
             if (!req.__nextReduxWrapperStore) {
-                req.__nextReduxWrapperStore = createStore();
+                req.__nextReduxWrapperStore = createStore(); // Used in GIP/GSSP
```
Turns out the server side hydration is in fact required. My idea of using a `serverStore` variable outside turned out to be a giant memory leak.

**IMPORTANT!**
Turns out that variables that you store in the module/global scope (like `serverStore`) **STAY ALIVE FOREVER**, that is: **CROSS REQUEST** (until you stop the server of course)! This means that the store that is saved in that variable will still be there when another **user** makes a request! Which means different users's client side stores will be hydrated with the previous user's store state. Bad bad bad 🤦‍♂️. TIL!

So we have to go back to using the req object to save the store in, which means we need to hydrate on the server when we reach the App component during server side rendering. In other words, I've reverted the change back to how it was before.

kirill-konshin replied:
I told you so...

voinik commented on lerna.json:
```diff
@@ -1,5 +1,5 @@
 {
-  "version": "7.0.0",
+  "version": "8.1.0",
```
I figured it's a good idea to bump the version to 8.1.0 as this PR changes when the hydration happens. I think the major version being 8 is needed because it's already version 8 on npm and in the releases view in GitHub. Thoughts?

kirill-konshin replied:
Pls revert this version change as version is determined by Lerna from CI git tags. Inside repo we can have like 1.0.0 everywhere ;)

voinik replied:
Done

kirill-konshin commented:
@voinik I will do one more round of review as soon as I can, thanks a lot for your help!

czekFREE commented:
@voinik, @kirill-konshin - I am playing with this and it looks good, but it seems that page-level getInitialProps is not triggering initial hydrate action (app-level getInitial props seems to be working fine).

Fix:
`  const useWrappedStore = (...): { ... } => {
    const phase = props?.__N_SSG ? 'gssg' : props?.__N_SSP ? 'gssp' : 'gip';

   ....

    useHybridHydrate(
      store,
      // This is the fix 
      phase === 'gip' ? initialState ?? props.pageProps.initialState : initialState,
      gspState,
      gsspState,
      phase
    );
    
    ...
`

In my code am also passing additional param "phase" to `hydrate` func to be able to better handle merging states based on when hydrate was dispatched - something that would be nice to have in master eventually as well...

voinik commented on packages/wrapper/src/index.tsx:
```diff
@@ -161,44 +161,69 @@ export const createWrapper = <S extends Store>(makeStore: MakeStore<S>, config:
-                hydrate(store, state);
-            }
-        }, [store, state]);
+            events?.on('routeChangeStart', handleStart);
```
I ran into an issue when doing `router.push` while staying on the same page. Hydrates didn't work anymore. This code makes sure hydrates run even if your router pushed while staying on the same page

kirill-konshin replied:
Good job here

kirill-konshin replied:
I was very skeptical about looking at URLs when I proposed that ;) it's better to listen to something more trustworthy.

kirill-konshin commented on .yarnrc.yml:
```diff
@@ -0,0 +1 @@
+nodeLinker: node-modules
```
Does everything still work with PNP?

voinik replied:
I'm not sure why this was added, although I can't run `yarn install` without it. Not sure what PNP is, maybe you mean PNPM? It works if you change the line in the `package.json` to this: 
```
"packageManager": "pnpm@7.18.2",
```

kirill-konshin replied:
PNP — Plug'n'play, yarn's linking method.

kirill-konshin commented:
I have finished the review, looks like everything is in order, except few nitpicks.

Main question is — are there any known problems?

voinik commented:
Hey @kirill-konshin thanks for the response!

> I have finished the review, looks like everything is in order, except few nitpicks.
> 
> Main question is — are there any known problems?

Great, thanks for reviewing! I've reverted the Lerna version change. 

As for issues: 
* The only one I could find was the router one that I fixed today.
* Apparently @czekFREE found an issue with `getInitialProps` usage in places **other** than _app. I could take a look at it but I think fixing that in another PR might be better as to not mess up anything this late into the PR development. Additionally, people should really be using `getServerSideProps` for that and not `getInitialProps` (docs support this). I'll leave it up to you to decide if I should fix it in this one
* The web app that I work on seems to not be having issues with this version (other than the router thing; I ran into that in the app I work on). It's a pretty big application and everything seems to be running smooth. It uses `getServerSideProps` as well as `getInitialProps` in _app. I haven't tested with `getStaticProps` much,  though. I've also asked a few Issue reporters to test this PR but none have reported back.

The only problem other than these is the one mentioned in the **caveats** section of Solution **B** in the PR description, namely the warning you get from React, but that's not visible in production.

Any thoughts on these points and anything I can do to make this PR ready for merge?

kirill-konshin commented on package.json:
```diff
@@ -1,7 +1,7 @@
   "name": "next-redux-wrapper-mono",
   "private": true,
-  "version": "7.0.0",
+  "version": "8.1.0",
```
You can revert all version numbers to 7.0.0. Lerna will do the trick.

kirill-konshin commented:
> Apparently @czekFREE found an issue with getInitialProps usage in places other than _app. I could take a look at it but I think fixing that in another PR might be better as to not mess up anything this late into the PR development.

Can you shed more light on this one?

Once you done with the versions in package.jsons, I can merge this PR in some branch (not master) and we can add more fixes on top of it.

kirill-konshin commented:
@voinik merged! Thanks a lot for your contribution. If there's anything else you want to fix — send PR to https://github.com/kirill-konshin/next-redux-wrapper/tree/async-hydration-fix-proper

voinik commented:
@kirill-konshin Alright I've reverted all version bumps. As for if PNP still works: No idea how to test that as I've never used yarn until I started working on this repo. Perhaps you can test it?

Regarding @czekFREE 's issue: I've just tested it and he is correct. I'm about to push my fix for it, just testing it a few more times to make sure it works. I've adjusted the redux-toolkit repo so it can be manually tested. I'll report back ASAP

voinik commented:
Haha I see you've already merged it. I'll send a PR to that branch! And my pleasure :)

saschati commented:
Hi, I have been using this change in my project for a few days, the project is big and we have integrated it with the 8.0.0 version of this library, after a long time I noticed the problem described in [#495](https://github.com/kirill-konshin/next-redux-wrapper/issues/495) and then I came across this request, I can say that after the changes in this request pool the site visually started behaving strangely, for a while and it's about half a second before the route change, the page loses its original visual appearance because it no longer has a given state, but the next one hasn't loaded yet, is there a way to fix it, except that at the beginning of changing the router to the content part, hang some kind of preloader so that it is not visible?

karaoak commented:
> 

thiis can depend how you have written your hydrate reducers. We encountered the issue you descibed only on one page, but fixed it by changing the reducer (not throwing away state directly) for the page you're navigating away from.

Avcajaraville commented:
@voinik I just tried this new release and it is working like a charm! Thanks a lot for your work <3
